src/dag.rs

```diff
@@ -0,0 +1,483 @@
+    /// a sequence of steps that run one after another.)
+    ///
+    /// `action` will be used when this node is being executed.
+    ///
+    /// The node is called `name`. This name is used along with `instance_id`
+    /// for storing the output of the action so that descendant nodes can
+    /// access it using [`crate::ActionContext::lookup`].
+    ///
+    /// The instance id is useful because there may be many subsagas of the same
+    /// type that run as part of the main saga. Each one of nodes in each subsaga
+    /// will only want to access data from its own nodes, and a common name for those nodes
+    /// is not enough to distinguish them across sagas. Furthermore, instance ids allow
+    /// the parent saga to identify data generated from different subsagas.
+    pub fn append(&mut self, node: Node) {
+        let newnode = self.graph.add_node(node);
+        for node in &self.last_added {
+            self.graph.add_edge(*node, newnode, ());
+        }
+
+        self.last_added = vec![newnode];
+    }
+
+    /// Adds a set of nodes to the graph that can be executed concurrently
+    ///
+    /// The new nodes will individually depend on completion of all actions that
+    /// were added in the last call to `append` or `append_parallel`.  `actions`
+    /// is a vector of `(name, action)` tuples analogous to the arguments to
+    /// [`SagaTemplateBuilder::append`].
+    pub fn append_parallel(&mut self, nodes: Vec<Node>) {
+        let newnodes: Vec<NodeIndex> =
+            nodes.into_iter().map(|node| self.graph.add_node(node)).collect();
+
+        // TODO-design For this exploration, we assume that any nodes appended
+        // after a parallel set are intended to depend on _all_ nodes in the
+        // parallel set.  This doesn't have to be the case in general, but if
+        // you wanted to do something else, you probably would need pretty
+        // fine-grained control over the construction of the graph.  This is
+        // mostly a question of how to express the construction of the graph,
+        // not the graph itself nor how it gets processed, so let's defer for
+        // now.
+        //
+        // Even given that, it might make more sense to implement this by
+        // creating an intermediate node that all the parallel nodes have edges
+        // to, and then edges from this intermediate node to the next set of
+        // parallel nodes.
+        for node in &self.last_added {
+            for newnode in &newnodes {
+                self.graph.add_edge(*node, *newnode, ());
+            }
+        }
+
+        self.last_added = newnodes;
+    }
+
+    /// Take a saga spec describing nodes, create the nodes,
+    /// and append them to the existing Dag.
+    ///
+    /// TODO: We probably want an `append_parallel_subsaga` that allows us to link all subsagas
+    /// to the `last_added` nodes in a parallel fashion.
+    pub fn append_subsaga<'a, T>(
```

Comment:
Should this instead be called `append_saga`? You can actually create the outer most saga with a `SagaSpec`. Additionally, since all nodes in all sagas/subsagas are joined into a single DAG, there really is no differentiation between a saga and a subsaga.

Reply:
This is fixed with the merge of changes from @davepacheco. `Dag`s now compose.